### PR TITLE
added sink for csv logging for cost/eval functions

### LINKED ISSUE #1842: Add Log level and Logging functions for cost/evaluation functions
### Description of the task

For debugging modules that rely on cost or evaluation functions we want to use the logger to generate a .csv file for analysis

### Acceptance criteria

- [ ] Add cost/evaluation log level to `custom_logging_level.h`
- [ ] Add cost/evaluation logging function to the custom sink to write values to a .csv file so that the .csv file can be plotted
- [ ] Modify the logging call function so that logs at this log level arent sent to the .log file

### Blocked By

#1754

### PR DESCRIPTION
**Pull request by LiCody**
### Description

- added log level for csv logging
- added sink for csv logging
- adjusted macro to support LOG(CSV, "filename")
- added test for the new sink

### Testing Done
Tests

### Resolved Issues

resolves #1842 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/logger/logger.h [hunk 1]
```diff
@@ -9,17 +9,34 @@
...
+#define LOG_SELECT(_1, _2, NAME, ...) NAME
+
+#undef LOG
+#define LOG(...) LOG_SELECT(__VA_ARGS__, LOG_2, LOG_1)(__VA_ARGS__)
+#define LOG_1(level)                                                                     \
+    if (!g3::logLevel(level))                                                            \
+    {                                                                                    \
+    }                                                                                    \
+    else                                                                                 \
+        INTERNAL_LOG_MESSAGE(level).stream()
```
Comment (jonathanlew) [inline src/software/logger/logger.h:30]: can we put curly braces after the else for clarity?

File: src/software/logger/logger.h [hunk 2]
```diff
@@ -9,17 +9,34 @@
...
+#define LOG_2(level, filename)                                                           \
+    if (level != CSV)                                                                    \
+    {                                                                                    \
+    }                                                                                    \
+    else                                                                                 \
+        LOG_1(level) << filename                                                         \
```
Comment (jonathanlew) [inline src/software/logger/logger.h:38]: same here

File: src/software/logger/logger.h [hunk 3]
```diff
@@ -9,17 +9,34 @@
...
- * the logWorker object must stay in scope, otherwise the logger is automatically
- * destroyed. So if an "init" function is used, the logWorker goes out of scope as
- * soon as the function returns, which destroys the logger right after creating it
- *
- * The Singleton class allows us to keep the logWorker in scope for the duration
- * of the program while still providing a single function to initialize the logger
+#define LOG_SELECT(_1, _2, NAME, ...) NAME
+
+#undef LOG
+#define LOG(...) LOG_SELECT(__VA_ARGS__, LOG_2, LOG_1)(__VA_ARGS__)
```
Comment (jonathanlew) [inline src/software/logger/logger.h]: can we have some comments explaining how this works?

File: src/software/logger/csv_sink.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            std::string file_name = msg.substr(0, pos);
+            std::string file_data = msg.substr(pos, msg.length());
+            std::ofstream csv_file(log_directory + "/" + file_name,
+                                   std::ios::out | std::ios_base::app);
+            csv_file << file_data;
+            //            std::cout << csv_file.good() << std::endl;
+            //            std::cout << csv_file.eof() << std::endl;
+            //            std::cout << csv_file.fail() << std::endl;
+            //            std::cout << csv_file.bad() << std::endl;
```
Comment (jonathanlew) [inline src/software/logger/csv_sink.cpp]: looks like extra debug couts

File: src/software/logger/csv_sink_test.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+class CSVSinkTest : public testing::TestWithParam<std::tuple<LEVELS>>
+{
+};
+const std::string logging_dir = "software/logger";
+TEST(CSVSinkTest, test_csv_log_appends)
+{
+    std::unique_ptr<g3::LogWorker> logWorker = g3::LogWorker::createLogWorker();
+    auto csv_sink_handle = logWorker->addSink(std::make_unique<CSVSink>(logging_dir),
+                                              &CSVSink::appendToFile);
```
Comment (jonathanlew) [inline src/software/logger/csv_sink_test.cpp:18]: this is what we'd have to add a binary to make it output to a custom directory?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nice work!

File: src/software/logger/logger.h (hunk 3 above)
Comment (LiCody) [reply src/software/logger/logger.h]: added

File: src/software/logger/logger.h (hunk 1 above)
Comment (LiCody) [reply src/software/logger/logger.h:30]: we end up with syntax errors because we need to pipe log messages

**Review by akhilveeraghanta — APPROVED** [review]
lgtm!

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**